docs/eager.md

```diff
@@ -0,0 +1,122 @@
+
+
+## Background
+Currently PyTorch/XLA runs on the LazyTensor tracing mode by default. In the following code
+```python
+import torch
+import torch_xla
+import torchvision
+
+device = torch_xla.device()
+model = torchvision.models.resnet18().to(device)
+input = torch.randn(64, 3, 224, 224).to(device)
+
+# model tracing
+res = model(input)
+
+# model execution, you can also use `xm.mark_step`
+torch_xla.sync()
+```
+The actual model compilation and device execution happens when `torch_xla.sync` is called. There are multiple drawback of this approach. 
+
+1. Users are often confused about when the framework is tracing and when the framework is executing.
+2. Non-core model code(data preprocessing for example) often generates some small pending execution that gets leaked into the main graph(step function) and causes recompilation. The recompilation of the whole graph is usually very expensive.
+3. It is hard to debug when/why recompilation happens.
+
+To mitigate above issues we want to introduce the new UX with eager and compile.
+
+## Basic Usage
+```python
+import torch
+import torch_xla
+import torchvision
+
+# Run ops eagerly by default
+torch_xla.experimental.eager_mode(True)
+
+device = torch_xla.device()
+model = torchvision.models.resnet18().to(device)
+
+# Mark the function to be compiled
+compiled_model = torch_xla.experimental.compile(model)
+input = torch.randn(64, 3, 224, 224).to(device)
+
+# Compilation and execution happens right away.
+res = compiled_model(input)
+```
+Note that
+
+1. Currently user has to manually enable the eager mode by `torch_xla.experimental.eager_mode(True)`.
+2. The region of the code that wants to be compiled should be wrapped by `torch_xla.experimental.compile`.
+
+The implementation of the `torch_xla.experimental.compile` is actually pretty straight forward, it disable the eager mode when entering the target function and start tracing. It will call the `torch_xla.sync()` when target function returns and reenable the eager mode. You can expect the same perfomrance by using the `eager` + `compile` API compared to the existing `mark_step/sync` approach.
+
+
+### Infernce
+```python
+torch_xla.experimental.eager_mode(True)
+
+compiled_model = torch_xla.compile(model, backend="openxla")
+```
+It is recommened to use the `torch.compile` instead of `torch_xla.experimental.compile` for inference to reduce the tracing overhad. 
+
+### Training
+```python
+torch_xla.experimental.eager_mode(True)
+
+def step_fn(self, data, target):
+    self.optimizer.zero_grad()
+    logits = self.model(data)
+    loss = self.loss_fn(
+        logits.view(-1, self.config.vocab_size), target.view(-1))
+    loss.backward()
+    optimizer.step()
+    return loss
+
+self.compiled_step_fn = torch_xla.experimental.compile(self.step_fn)
```

Comment:
re: naming, I don't think `compiled_step_fn` makes sense here. It makes it look like this function is imperatively compiled when you call `compile`, when what's really happening is that you will trace and JIT-compile the function when it's called later. `jit` and `trace` already mean something else in this ecosystem though...

Reply:
well that's the trade of I made when I try to align the api with the upstream, I think it creates more confusion to called it `jited_step_fn`. I will just rename it to `step_fn`

Reply:
Do you want to just use `compile` as a decorator here if you're following the upstream convention?

Reply:
seem like upstream support both, I can support both too https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html